### PR TITLE
Create setData() function on nimble-table and remove data property

### LINKED ISSUE #958: nimble-table shouldn't require clients to shallow copy their data
# 🐛 Bug Report

See [this discussion from a PR](https://github.com/ni/nimble/pull/953#discussion_r1067525680).

The summary is:
- We should change `data` on the table to have a setter rather than rely on FAST's `@observable` change detection so that we can ensure the table updates in more scenarios
- Once we have a setter for `data`, we should do a shallow copy of the data before passing it to TanStack to ensure TanStack detects the update

## 🤔 Expected Behavior

The client of the table should be able to update their data like this:
```TS
myData = [ /* some data here */ ];
myTable.data = myData;

myData[0].myProperty = 'new-value';
// Ideally, this should update the first row in the table with 'new-value'
// value for 'myProperty', but it doesn't actually update the table's rendering
myTable.data = myData;

// must do this instead
myTable.data = [ ...myData ];
```

## 😯 Current Behavior

In the example above, the client must do a shallow copy of their data before reassigning it to the table -- shown in the last line:
```
myTable.data = [ ...myData ];
```

## 💁 Possible Solution

See notes in description

## 🌍 Your Environment

Not environment specific

### PR DESCRIPTION
**Pull request by mollykreis**
# Pull Request

## 🤨 Rationale

Fixes #958 

The current `data` property on the `nimble-table` relies on FAST's `@observable` behavior for arrays to detect updates. This can cause some unexpected behavior for clients, including having to make a shallow copy of their data array in some cases. To avoid these oddities and to keep the `Table` class from having to hold on to its own copy of the data, the `data` property is being removed and replaced with a `setData` function that always performs a shallow copy of the data and reassigns it to TanStack.

## 👩‍💻 Implementation

Web Component
- Remove `data` property
- Add `setData()` function that contains much of the logic the removed `dataChanged()` function performed
- Perform a shallow copy of each row's data when creating `TableRowState` instances to ensure that the `@observable` handler for the `TableRow` was executed when `dataRecord` changes.
- Introduced a `canRenderRows` property that tracks whether the rows should be rendered in the table based on the table's validity. This allowed the table code to always update TanStack with the data (previously it gave TanStack an empty array when the data was invalid) and avoid having to keep a separate copy of the data within the `Table` class.

Angular
- Remove `data` property
- Add `setData()` function
- Add `data$` property that encapsulates the logic to call `setData()` when an Observable emits a new value

Blazor
- Update the interop code to call `setData()` on the web component rather than assigning to the `data` property
- Did not expose a `setData()` method because it would only wrap a line of code assigning to the Blazor Table's `Data` property

## 🧪 Testing

- Created auto tests for component and Angular changes
- Manually tested Angular and Blazor example apps
- Updated matrix tests to have a play function that adds data to the table

## ✅ Checklist

- [ ] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

File: change/@ni-nimble-components-af276551-d3c5-47e9-8241-5a435257aa31.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (jattasNI) [inline change/@ni-nimble-components-af276551-d3c5-47e9-8241-5a435257aa31.json:1]: Could you update the HLD where we describe the `data` property (and our rationale for initially choosing to use a property instead of a method)?

File: packages/nimble-components/src/table/tests/table.stories.ts [hunk 2]
```diff
@@ -48,10 +49,19 @@ const dataSetIdFieldNames = {
...
 const overviewText = 'The `nimble-table` is a component that offers a way to render tabular data in a variety of ways in each column.';
 
-const dataDescription = `\`data\` is a property that is an array of records. A record provides the data that backs a single row in the table.
-Each record is made up of fields, which are key/value pairs. The key in each pair must be of type \`string\`, which is defined by the type
-\`TableFieldName\`. The value in each pair must be of type \`string\`, \`number\`, \`boolean\`, \`Date\`, \`null\`, or \`undefined\`,
-which is defined by the type \`TableFieldValue\`.`;
+const dataDescription = `To set the data on the table, call \`setData()\` with an array data records. Each record is made up of fields,
+which are key/value pairs. The key in each pair must be of type \`string\`, which is defined by the type \`TableFieldName\`. The value
+in each pair must be of type \`string\`, \`number\`, \`boolean\`, \`Date\`, \`null\`, or \`undefined\`, which is defined by the type \`TableFieldValue\`.
+
```
Comment (jattasNI) [inline packages/nimble-components/src/table/tests/table.stories.ts:55]: I'd vote to be really explicit about when it's necessary to call `setData()` if a client wants to update the table. Something like this maybe (assuming my understanding is correct)?

```suggestion
The table will not automatically update if the contents of the array change after calling `setData()`. To trigger an update you can call `setData()` again with the same array reference.
```

File: packages/nimble-components/src/table/tests/table.spec.ts [hunk 3]
```diff
@@ -161,48 +161,83 @@ describe('Table', () => {
...
             simpleTableData[0]
         ];
-        element.data = updatedData;
+        element.setData(updatedData);
         await waitForUpdatesAsync();
 
-        verifyRenderedData();
+        verifyRenderedData(updatedData);
+    });
+
```
Comment (jattasNI) [inline packages/nimble-components/src/table/tests/table.spec.ts:202]: Could you add a test that verifies the table **doesn't** update if the original data array is changed after passing it to `setData()`? Basically ensuring the table doesn't hold on to a reference to the original array.

Comment (jattasNI) [inline packages/nimble-components/src/table/tests/table.spec.ts:202]: Could you add a test that calls `setData()` with an empty array? There's some conditional code in the implementation so it would be good to verify what happens if it's not called.

**Review by jattasNI — APPROVED** [review]

File: packages/nimble-components/src/table/tests/table.stories.ts [hunk 4]
```diff
@@ -48,10 +49,22 @@ const dataSetIdFieldNames = {
...
+The table will not automatically update if the contents of the array change after calling \`setData()\`. To trigger an update, call
+\`setData()\` again with the same array reference or with a new array.
+
+<details>
+    <summary>Framework specific considerations</summary>
+    - Angular: In addition to exposing the \`setData()\` function in Angular, you can use the \`data$\` property to provide an
+    \`Observable<TableRecord[]>\`. Nimble will automatically subscribe and unsubscribe to the provided \`Observable\` and call
+    \`setData()\` on the web component when new values are emitted.
+    - Blazor: In addition to exposing the \`setData\` function in Blazor, you can also use the \`Data\` property on the
```
Comment (atmgrifter00) [inline packages/nimble-components/src/table/tests/table.stories.ts:64]: This isn't accurate right? We aren't exposing the `setData` function in Blazor at all.

**Review by atmgrifter00 — APPROVED** [review]

File: packages/nimble-components/src/table/tests/table.stories.ts (hunk 4 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/tests/table.stories.ts:64]: Yes, you're right. Good catch!

**Review by msmithNI — APPROVED** [review]

File: packages/nimble-components/src/table/tests/table-matrix.stories.ts [hunk 5]
```diff
@@ -57,6 +58,12 @@ export const tableThemeMatrix: Story = createMatrixThemeStory(
     createMatrix(component)
 );
 
+tableThemeMatrix.play = (): void => {
```
Comment (jattasNI) [inline packages/nimble-components/src/table/tests/table-matrix.stories.ts:61]: >Updated storybook story for the table to have a button that loads rows into the table
>
>    Ideally, I would have liked to have this done automatically, but I couldn't find a way to do this because the play function was only called when the story initially loads, not when args are changed.

An unfortunate side effect of this is that the Chromatic snapshot for the non-matrix story is now an empty table. At least we still have test coverage for row/cell appearance via the matrix story, but it would be nicer if we could have the table be populated with data initially.

Could you describe the behavior in more detail if we add a `play` function which calls `setData()` to the non-matrix story? Would it start out with data (including in the Chromatic snapshot) but then clear it whenever the user changes an argument? Or does Storybook do something on load that causes the data to never be populated?

Comment (mollykreis) [reply packages/nimble-components/src/table/tests/table-matrix.stories.ts:61]: We could implement `play` to call `setData()` on the non-matrix story. When you change anything in the control panel of storybook (which cases the table to re-render), the table data would be cleared and `play` would not be called again.

File: packages/nimble-components/src/table/index.ts [hunk 6]
```diff
@@ -112,7 +106,7 @@ export class Table<
         const rows = this.table.getRowModel().rows;
         this.tableData = rows.map(row => {
             const rowState: TableRowState<TData> = {
-                record: row.original,
+                record: { ...row.original },
```
Comment (rajsite) [inline packages/nimble-components/src/table/index.ts:109]: hmmmmm this looks like it moves us from holding a reference to the original record object to make a copy. And it seems like this runs for all rows in the table. Is that the intention (i.e. so we are not holding onto references in the original object)?

If so, how often is refreshRows called? If it is more often than setData then maybe instead we should do a full copy once in setData instead and then here continue to just pass a reference since we own the full copy at that point?

Maybe we should just do that copy in setData instead so it's very clear where the cost of a copy is being made.

Also dates are unfortunately mutable objects, so for a proper clone we actually need to make a copy, ie `new Date(originalDate)`. 

I hadn't considered that before but this actually might be a useful change. The Date object instance really only has one piece of state associated with it, an number which is milliseconds since the unix epoch. Maybe the table should not take Date objects but instead take the number representation (ie `date.valueOf()` which is equal to `date.getTime()`);. If we make the user pass the number instead then we can make performant copies without concern like with the other primitive types. I don't think that's a huge amount of overhead for users either. Alternatively in our setData copy we could do that conversion for them.

File: packages/nimble-components/src/table/tests/table.stories.ts [hunk 7]
```diff
@@ -93,7 +106,13 @@ const metadata: Meta<TableArgs> = {
             WARNING - The table is still in development and considered
             experimental. It is not recommended for application use.
         </div>
-        <nimble-table :data=${x => dataSets[x.data]} id-field-name=${x => dataSetIdFieldNames[x.data]}>
+        <nimble-button @click="${x => x.populateData(x.tableRef, x.data)}">
+            Load data
+        </nimble-button>
+        <nimble-table
+            ${ref('tableRef')}
+            id-field-name="${x => dataSetIdFieldNames[x.data]}"
```
Comment (rajsite) [inline packages/nimble-components/src/table/tests/table.stories.ts:114]: Doing something like this worked for me locally to bind to the data set. Just abusing the binding a bit but then don't need the button

image

File: packages/nimble-components/src/table/tests/table-matrix.stories.ts (hunk 5 above)
Comment (jattasNI) [reply packages/nimble-components/src/table/tests/table-matrix.stories.ts:61]: Sounds like Milan's suggestion below might remove the need for a button. 

If that doesn't work for some reason then I'm actually now leaning back in the direction you initially proposed of not having a play function at all. Chromatic test coverage of the non-matrix story isn't that important and I think the user experience is more consistent if they learn to click the button from the very beginning.

File: packages/nimble-components/src/table/index.ts [hunk 8]
```diff
@@ -63,7 +65,14 @@ export class Table<
             autoResetAll: false
         };
         this.table = tanStackCreateTable(this.options);
-        this.tableInitialized = true;
+    }
+
+    public setData(newData: readonly TData[]): void {
+        if (newData.length > 0 && this.table.options.columns.length === 0) {
```
[the remark this answers is not in the document]
Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts:71]: We create TanStack columns to correspond to the fields in `TData`, not the visible columns that are configured in our table. Unless we expect the fields in `TData` to change when `setData()` is called, we don't need to update TanStack's columns.

[the remark this answers is not in the document]
Comment (rajsite) [reply packages/nimble-components/src/table/index.ts:71]: > Unless we expect the fields in TData to change when setData() is called

I may be misunderstanding, but yea sure, why not? Call setData with different fields. It'll be in an invalid config and not render until you also update the column configuration field names, but tanstack doesn't care about that. It just cares about aligning with TData.

```ts
table.setData([{a:1, b:1, c:1}]);
table.setData([{e:1, f:1, g:1}]); // The fields in TData are changed when setData is called and we need to update TanStack columns
// table not rendering until the user also goes in and updates column configuration
```

File: packages/nimble-components/src/table/components/row/index.ts [hunk 9]
```diff
@@ -25,6 +25,9 @@ export class TableRow<
     @attr({ attribute: 'record-id' })
     public recordId?: string;
 
+    // The template doesn't bind to `dataRecord`, but making it observable
+    // causes all the template bindings to be re-evaluated when the value changes
+    // because it is part of the binding context for the template.
```
[the remark this answers is not in the document]
Comment (mollykreis) [reply packages/nimble-components/src/table/components/row/index.ts]: The issue with having `cellState` be an accessor on the cell is that something would still need to tell cell that the row's record had been updated.

The approach that I went with is:
- Get rid of `getCellState()` on the row
- Create a `@volatile` property on the row named `columnStates` that returns an array of column + cell state. The implementation of this property is very similar to the implementation of `getCellState()`, but it creates the states for all columns
- Bind the row's template to the new `columnStates` property instead of to the `columns` property

**Review by rajsite — APPROVED** [review]

File: packages/nimble-components/src/table/components/row/index.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (rajsite) [reply packages/nimble-components/src/table/components/row/index.ts]: lgtm 🚢-it!
